xds/src/main/java/io/grpc/xds/RingHashLoadBalancer.java

```diff
@@ -93,81 +95,198 @@ public boolean acceptResolvedAddresses(ResolvedAddresses resolvedAddresses) {
+          (double) config.maxRingSize);
+
+      // Build the ring
+      ring = buildRing(serverWeights, totalWeight, scale);
+
+      // If we need to proactively start connecting, iterate through all the subchannels, starting
+      // at a random position.
+      // Alternatively, we should better start at the same position.
+      Collection<ChildLbState> childLbStates = getChildLbStates();
+      connectionAttemptIterator = childLbStates.iterator();
+      int randomAdvance = random.nextInt(childLbStates.size());
+      while (randomAdvance-- > 0) {
+        connectionAttemptIterator.next();
+      }
+
+      // Must update channel picker before return so that new RPCs will not be routed to deleted
+      // clusters and resolver can remove them in service config.
+      updateOverallBalancingState();
+
+      shutdownRemoved(acceptRetVal.removedChildren);
+    } finally {
+      this.resolvingAddresses = false;
     }
-    long minWeight = Collections.min(serverWeights.values());
-    double normalizedMinWeight = (double) minWeight / totalWeight;
-    // Scale up the number of hashes per host such that the least-weighted host gets a whole
-    // number of hashes on the the ring. Other hosts might not end up with whole numbers, and
-    // that's fine (the ring-building algorithm can handle this). This preserves the original
-    // implementation's behavior: when weights aren't provided, all hosts should get an equal
-    // number of hashes. In the case where this number exceeds the max_ring_size, it's scaled
-    // back down to fit.
-    double scale = Math.min(
-        Math.ceil(normalizedMinWeight * config.minRingSize) / normalizedMinWeight,
-        (double) config.maxRingSize);
-    ring = buildRing(serverWeights, totalWeight, scale);
-
-    // Shut down subchannels for delisted addresses.
-    List<Subchannel> removedSubchannels = new ArrayList<>();
-    for (EquivalentAddressGroup addr : removedAddrs) {
-      removedSubchannels.add(subchannels.remove(addr));
+
+    return true;
+  }
+
+  /**
+   * Updates the overall balancing state by aggregating the connectivity states of all subchannels.
+   *
+   * <p>Aggregation rules (in order of dominance):
+   * <ol>
+   *   <li>If there is at least one subchannel in READY state, overall state is READY</li>
+   *   <li>If there are <em>2 or more</em> subchannels in TRANSIENT_FAILURE, overall state is
+   *   TRANSIENT_FAILURE</li>
+   *   <li>If there is at least one subchannel in CONNECTING state, overall state is
+   *   CONNECTING</li>
+   *   <li> If there is one subchannel in TRANSIENT_FAILURE state and there is
+   *    more than one subchannel, report CONNECTING </li>
+   *   <li>If there is at least one subchannel in IDLE state, overall state is IDLE</li>
+   *   <li>Otherwise, overall state is TRANSIENT_FAILURE</li>
+   * </ol>
+   */
+  @Override
+  protected void updateOverallBalancingState() {
```

Comment:
This override breaks sticky-TF.  I think ring hash needs its own sticky TF, regardless of pick first lb's sticky TF.

Reply:
> Only PF needs to deal with sticky TF. The others will only ever go from TF to something other than READY when a new address is added, and in that case you want to go to IDLE/CONNECTING.

That makes sense.

Note that in the current new Pick First,  if the current state is in TF, the next state that will be reported is either READY and TF. We report TF again just because of notifying the new error message in TF. It was unclear at that time whether it worth the notification overhead and also because we thought petiole would have its own sticky TF. Since that is changed, the decision in the new pick first needs to be revised.

Reply:
https://github.com/grpc/grpc-java/blob/master/core/src/main/java/io/grpc/internal/PickFirstLeafLoadBalancer.java#L238

Reply:
I don't see what needs to change.  That line says that we would report TF again which seems appropriate. We would want the ChildLbState.currentPicker to be updated with the new description.  The petiole policy not doing sticky TF is irrelevant as far as I can see.

Reply:
Right, probably nothing needs to be changed. Sticky TF should allow transition to TF from TF. Glad that we verified that.